File: frontend/src/utils/standings.ts
import type { Match, StandingRow, Team } from '../types';

export function calculateStandings(teams: Team[], matches: Match[]): StandingRow[] {
  const finished = matches.filter((m) => m.status === 'finished');
  const stats = new Map<
    string,
    Omit<StandingRow, 'position' | 'teamName'>
  >();

  for (const team of teams) {
    stats.set(team.id, {
      teamId: team.id,
      played: 0,
      wins: 0,
      draws: 0,
      losses: 0,
      goalsFor: 0,
      goalsAgainst: 0,
      goalDifference: 0,
      points: 0,
    });
  }

  for (const match of finished) {
    const home = stats.get(match.homeTeamId);
    const away = stats.get(match.awayTeamId);
    if (!home || !away) continue;

    home.played += 1;
    away.played += 1;
    home.goalsFor += match.homeScore;
    home.goalsAgainst += match.awayScore;
    away.goalsFor += match.awayScore;
    away.goalsAgainst += match.homeScore;

    if (match.homeScore > match.awayScore) {
      home.wins += 1;
      home.points += 3;
      away.losses += 1;
    } else if (match.homeScore < match.awayScore) {
      away.wins += 1;
      away.points += 3;
      home.losses += 1;
    } else {
      home.draws += 1;
      away.draws += 1;
      home.points += 1;
      away.points += 1;
    }
  }

  const teamNames = new Map(teams.map((t) => [t.id, t.name]));

  return [...stats.values()]
    .map((row) => ({
      ...row,
      teamName: teamNames.get(row.teamId) ?? 'Unknown',
      goalDifference: row.goalsFor - row.goalsAgainst,
    }))
    .sort((a, b) => {
      if (b.points !== a.points) return b.points - a.points;
      if (b.goalDifference !== a.goalDifference) return b.goalDifference - a.goalDifference;
      return b.goalsFor - a.goalsFor;
    })
    .map((row, index) => ({ ...row, position: index + 1 }));
}

File: backend/app/store/standings.py
from app.models.schemas import Match, MatchStatus, StandingRow, Team


def calculate_standings(teams: list[Team], matches: list[Match]) -> list[StandingRow]:
    finished = [m for m in matches if m.status == MatchStatus.FINISHED]
    stats: dict[str, dict] = {}

    for team in teams:
        stats[team.id] = {
            "team_id": team.id,
            "played": 0,
            "wins": 0,
            "draws": 0,
            "losses": 0,
            "goals_for": 0,
            "goals_against": 0,
            "points": 0,
        }

    for match in finished:
        home = stats.get(match.home_team_id)
        away = stats.get(match.away_team_id)
        if home is None or away is None:
            continue

        home["played"] += 1
        away["played"] += 1
        home["goals_for"] += match.home_score
        home["goals_against"] += match.away_score
        away["goals_for"] += match.away_score
        away["goals_against"] += match.home_score

        if match.home_score > match.away_score:
            home["wins"] += 1
            home["points"] += 3
            away["losses"] += 1
        elif match.home_score < match.away_score:
            away["wins"] += 1
            away["points"] += 3
            home["losses"] += 1
        else:
            home["draws"] += 1
            away["draws"] += 1
            home["points"] += 1
            away["points"] += 1

    team_names = {t.id: t.name for t in teams}
    rows = []
    for row in stats.values():
        gd = row["goals_for"] - row["goals_against"]
        rows.append(
            StandingRow(
                position=0,
                team_id=row["team_id"],
                team_name=team_names.get(row["team_id"], "Unknown"),
                played=row["played"],
                wins=row["wins"],
                draws=row["draws"],
                losses=row["losses"],
                goals_for=row["goals_for"],
                goals_against=row["goals_against"],
                goal_difference=gd,
                points=row["points"],
            )
        )

    rows.sort(key=lambda r: (-r.points, -r.goal_difference, -r.goals_for))
    return [
        row.model_copy(update={"position": index + 1}) for index, row in enumerate(rows)
    ]
